core/src/test/java/org/elasticsearch/threadpool/FixedThreadPoolTests.java

```diff
@@ -0,0 +1,106 @@
+
+import java.util.concurrent.CountDownLatch;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+
+public class FixedThreadPoolTests extends ESThreadPoolTestCase {
+
+    public void testRejectedExecutionCounter() throws InterruptedException {
+        final String threadPoolName = randomThreadPool(ThreadPool.ThreadPoolType.FIXED);
+        // some of the fixed thread pool are bound by the number of
+        // cores so we can not exceed that
+        final int size = randomIntBetween(1, EsExecutors.boundedNumberOfProcessors(Settings.EMPTY));
+        final int queueSize = randomIntBetween(1, 16);
+        final long rejections = randomIntBetween(1, 16);
+
+        ThreadPool threadPool = null;
+        final Settings nodeSettings =
+            Settings.builder()
+                .put("node.name", "testRejectedExecutionCounter")
+                .put("threadpool." + threadPoolName + ".size", size)
+                .put("threadpool." + threadPoolName + ".queue_size", queueSize)
+                .build();
+        try {
+            threadPool = new ThreadPool(nodeSettings);
+
+            // these tasks will consume the thread pool causing further
+            // submissions to queue
+            final CountDownLatch latch = new CountDownLatch(size);
+            final CountDownLatch block = new CountDownLatch(1);
+            for (int i = 0; i < size; i++) {
+                threadPool.executor(threadPoolName).execute(() -> {
+                    try {
+                        latch.countDown();
+                        block.await();
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+            }
+
+            // wait for the submitted tasks to be consumed by the thread
+            // pool
+            latch.await();
+
+            // these tasks will fill the thread pool queue
+            for (int i = 0; i < queueSize; i++) {
+                threadPool.executor(threadPoolName).execute(() -> {});
+            }
+
+            // these tasks will be rejected
+            long counter = 0;
+            for (int i = 0; i < rejections; i++) {
+                try {
+                    threadPool.executor(threadPoolName).execute(() -> {});
+                } catch (EsRejectedExecutionException e) {
+                    counter++;
+                }
+            }
+
+            block.countDown();
+
+            assertThat(counter, equalTo(rejections));
+            assertThat(stats(threadPool, threadPoolName).getRejected(), equalTo(rejections));
+
+            // the rejected execution count resets to zero when the
+            // queue is resized
+            final ClusterSettings clusterSettings =
+                new ClusterSettings(nodeSettings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+            threadPool.setClusterSettings(clusterSettings);
+            clusterSettings.applySettings(
+                Settings.builder()
```

Comment:
This might not actually change the setting. Does that matter? Maybe set it to `queueSize + 1`?